dipy/segment/search.py

```diff
@@ -0,0 +1,292 @@
+                                np.all(barycenters <= self.max_box, axis=1))
+
+        baryc_to_box = barycenters[in_bin] - self.min_box
+        baryc_bins_id = (baryc_to_box // self.bin_size).astype(int)
+        baryc_multiid = np.ravel_multi_index(baryc_bins_id.T, self.bin_shape)
+
+        sort_id = np.argsort(baryc_multiid)
+        u_bin, mapping = np.unique(baryc_multiid[sort_id], return_index=True)
+        slines_ids = np.split(np.flatnonzero(in_bin)[sort_id], mapping[1:])
+        return u_bin, slines_ids
+
+
+def nearest_from_matrix_row(coo_matrix):
+    """
+    Return the nearest (smallest) for each row given an coo sparse matrix
+
+    Parameters
+    ----------
+    coo_matrix : scipy COOrdinates sparse matrix (nb_slines x nb_slines_ref)
+        Adjacency matrix containing all neighbors within the given radius
+
+    Returns
+    -------
+    non_zero_ids : numpy array (nb_non_empty_row x 1)
+        Indices of each non-empty slines (row)
+    nearest_id : numpy array (nb_non_empty_row x 1)
+        Indices of the nearest reference match (column)
+    nearest_dist : numpy array (nb_non_empty_row x 1)
+        Distance for each nearest match
+    """
+    non_zero_ids = np.unique(coo_matrix.row)
+    sparse_matrix = np.abs(coo_matrix.tocsr())
+    upper_limit = np.max(sparse_matrix.data) + 1.0
+    sparse_matrix.data = upper_limit - sparse_matrix.data
+    nearest_id = np.squeeze(sparse_matrix.argmax(axis=1).data)[non_zero_ids]
+    nearest_dist = upper_limit - np.squeeze(sparse_matrix.max(axis=1).data)
+    return non_zero_ids, nearest_id, nearest_dist
+
+
+def nearest_from_matrix_col(coo_matrix):
+    """
+    Return the nearest (smallest) for each column given an coo sparse matrix
+
+    Parameters
+    ----------
+    coo_matrix : scipy COOrdinates sparse matrix (nb_slines x nb_slines_ref)
+        Adjacency matrix containing all neighbors within the given radius
+
+    Returns
+    -------
+    non_zero_ids : numpy array (nb_non_empty_col x 1)
+        Indices of each non-empty reference (column)
+    nearest_id : numpy array (nb_non_empty_col x 1)
+        Indices of the nearest slines match (row)
+    nearest_dist : numpy array (nb_non_empty_col x 1)
+        Distance for each nearest match
+    """
+    non_zero_ids = np.unique(coo_matrix.col)
+    sparse_matrix = np.abs(coo_matrix.tocsc())
+    upper_limit = np.max(sparse_matrix.data) + 1.0
+    sparse_matrix.data = upper_limit - sparse_matrix.data
+    nearest_id = np.squeeze(sparse_matrix.argmax(axis=0).data)[non_zero_ids]
+    nearest_dist = upper_limit - np.squeeze(sparse_matrix.max(axis=0).data)
+    return non_zero_ids, nearest_id, nearest_dist
+
+
+def mean_l2_func(a, b):
+    """Streamlines distance function: Average L2 (MDF without flip)"""
+    return np.mean(np.sqrt(np.sum((a - b) ** 2, axis=-1)), axis=-1)
+
+
+def mean_l1_func(a, b):
+    """Streamlines distance function: Average L1"""
+    return np.mean(np.sum(np.abs(a - b), axis=-1), axis=-1)
```

Comment:
we might want to move these distances functions somewhere else..... I need to look where...

Reply:
moved into   dipy.segment.metric